feat: Adds 'skip-empty-xacts' support

adrijshikhar opened the pull request:
fixes #106

adrijshikhar committed:
feat: Adds 'skip-empty-xacts' support

adrijshikhar commented:
@eulerto Please have a look and let me know what changes are required

eulerto commented on README.md:
```diff
@@ -115,6 +115,7 @@ Parameters
 * `include-transaction`: emit records denoting the start and end of each transaction. Default is _true_.
 * `include-unchanged-toast` (deprecated): Don't use it. It is deprecated.
 * `filter-origins`: exclude changes from the specified origins. Default is empty which means that no origin will be filtered. It is a comma separated value.
+* `skip-empty-xacts` ignore empty transactions. Default is _false_.
```
do not include empty transactions.

adrijshikhar replied:
should i make it `true` by default?

eulerto replied:
No. It should be backward compatible.

eulerto commented on sql/skip_empty_xacts.sql:
```diff
@@ -0,0 +1,33 @@
+CREATE TABLE xact_test(data text);
+INSERT INTO xact_test VALUES ('before-test');
+
+-- bug #13844, xids in non-decoded records need to be inspected
```
#13844? There is no such bug report.

eulerto commented on sql/skip_empty_xacts.sql:
```diff
@@ -0,0 +1,33 @@
+-- and now show those changes
+SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1');
+
+-- bug #14279, do not propagate null snapshot from subtransaction
```
#14279? There is no such bug report.

eulerto commented on sql/skip_empty_xacts.sql:
```diff
@@ -0,0 +1,33 @@
+-- predictability
+SET synchronous_commit = on;
+
+SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2json');
```
Do these tests exercise the proposal? Why don't you use simple statements (for example, using `filter-tables` or `add-tables` option)? You should include tests for both formats (see other test files).

adrijshikhar replied:
ok

eulerto commented on wal2json.c:
```diff
@@ -95,6 +96,18 @@ typedef struct
 } JsonDecodingData;
 
+/*
+ * Maintain the per-transaction level variables to track whether the
```
Maintain a per-transaction structure...

eulerto commented on wal2json.c:
```diff
@@ -784,26 +814,48 @@ pg_filter_by_origin(LogicalDecodingContext *ctx, RepOriginId origin_id)
 static void
 pg_decode_begin_txn(LogicalDecodingContext *ctx, ReorderBufferTXN *txn)
+{
+	JsonDecodingData *data = ctx->output_plugin_private;
```
There are some space issues here.

adrijshikhar replied:
ok
how to fix that?
which IDE and linter you use?

eulerto replied:
Remove the spaces and check the diff again. I use vim. The rules for indentation comes from pgindent (see `src/tools/pgindent` in the Postgres tree).

adrijshikhar replied:
i tried `pgindent` but it made a lot of changes. i don't think I used the right `typedef` file

eulerto commented on wal2json.c:
```diff
@@ -784,26 +814,48 @@ pg_filter_by_origin(LogicalDecodingContext *ctx, RepOriginId origin_id)
+
+    txndata->xact_wrote_changes = false;
+    txn->output_plugin_private = txndata;
+
```
Remove this empty line.

eulerto commented on wal2json.c:
```diff
@@ -784,26 +814,48 @@ pg_filter_by_origin(LogicalDecodingContext *ctx, RepOriginId origin_id)
+    txn->output_plugin_private = txndata;
+
+
+	/*
```
There is some space issue here.

eulerto commented on wal2json.c:
```diff
@@ -784,26 +814,48 @@ pg_filter_by_origin(LogicalDecodingContext *ctx, RepOriginId origin_id)
+     * If asked to skip empty transactions, we'll emit BEGIN at the point
+     * where the first operation is received for this transaction.
+     */
+    if (data->skip_empty_xacts){
```
The style is to add the bracket in the next line. There are also some space issues here.

eulerto commented on wal2json.c:
```diff
@@ -784,26 +814,48 @@ pg_filter_by_origin(LogicalDecodingContext *ctx, RepOriginId origin_id)
+     * where the first operation is received for this transaction.
+     */
+    if (data->skip_empty_xacts){
+		elog(DEBUG2, "Skipping empty transaction");
```
Use lowercase here. "skipping an empty transaction"

eulerto commented on wal2json.c:
```diff
@@ -784,26 +814,48 @@ pg_filter_by_origin(LogicalDecodingContext *ctx, RepOriginId origin_id)
+}
+
+static void
+pg_output_begin(LogicalDecodingContext *ctx, ReorderBufferTXN *txn, bool last_write)
```
last_write: I don't like this variable name. Shouldn't it be called "wrote_change"?

adrijshikhar replied:
wrote_change seems better

eulerto commented on wal2json.c:
```diff
@@ -784,26 +814,48 @@ pg_filter_by_origin(LogicalDecodingContext *ctx, RepOriginId origin_id)
 
 	/* Transaction starts */
-	OutputPluginPrepareWrite(ctx, true);
+	OutputPluginPrepareWrite(ctx, last_write);
```
Did you test this modification with the `write-in-chunks` option?

adrijshikhar replied:
no i have not
i will benchmark this change with `write-in-chunks` option

eulerto commented on wal2json.c:
```diff
@@ -1578,6 +1638,14 @@ pg_decode_change(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
+
+	/* output BEGIN if we haven't yet */
+	if (data->skip_empty_xacts && !txndata->xact_wrote_changes)
+	{
```
You don't need the brackets here. It is just a single statement.

eulerto commented on wal2json.c:
```diff
@@ -2569,6 +2637,15 @@ static void pg_decode_truncate(LogicalDecodingContext *ctx,
+
+	/* output BEGIN if we haven't yet */
+	if (data->skip_empty_xacts && !txndata->xact_wrote_changes)
+	{
```
Same as the previous comment.

eulerto commented on wal2json.c:
```diff
@@ -170,7 +186,7 @@ static void pg_decode_truncate_v1(LogicalDecodingContext *ctx,
 /* version 2 */
 static void pg_decode_begin_txn_v2(LogicalDecodingContext *ctx,
-					ReorderBufferTXN *txn);
+					 ReorderBufferTXN *txn, bool last_write);
```
There is an additional space here.

eulerto commented on wal2json.c:
```diff
@@ -95,6 +96,18 @@ typedef struct
+ * has requested to skip the empty transactions we can skip the empty streams
+ * even though the transaction has written some changes.
+ */
+typedef struct
```
Why do you need this new struct? Can't you use `data->nr_changes`?

eulerto reviewed (commented):
Initial review.

adrijshikhar commented:
@eulerto can you please go through the comments I posted?
Thanks

eulerto commented:
I don't follow. You asked a question about `pgindent`. AFAICS you didn't post another update solving all concerns. I suggest that you read [Creating Clean Patches](https://wiki.postgresql.org/wiki/Creating_Clean_Patches) and [Development Tools and Help](https://wiki.postgresql.org/wiki/Developer_FAQ#Development_Tools_and_Help) (there is a subsection called _What's the formatting style used in PostgreSQL source code?_.

adrijshikhar commented:
thanks I will go through that. I have addressed all of your comments and made changes accordingly. I am stuck on `pgindent`. Somehow it changes all of the indentations. I will try to fix it and then update you/

eulerto commented:
If it is changing a lot, don't run it. Instead, fix some superfluous changes (such as whitespace and new line additions). I can clean it up later, if required.

adrijshikhar committed:
review and lint changes

adrijshikhar committed:
lint changes

adrijshikhar committed:
lint changes
